packages/arb-token-bridge-ui/src/components/TransactionHistory/TransactionsTableDetailsTeleporterSteps.tsx

```diff
@@ -0,0 +1,169 @@
+        1. To complete the deposit to{' '}
+        {getNetworkName(getChainIdForRedeemingRetryable(tx))}.{' '}
+        {!l1L2RetryableRequiresRedeem(tx) && (
+          <CheckIcon height={12} className="m-1 text-green-400" />
+        )}
+      </div>
+      <div className="flex items-center">
+        2. To relay the deposit to {getNetworkName(tx.destinationChainId)}.
+        {!l2ForwarderRetryableRequiresRedeem(tx) && (
+          <CheckIcon height={12} className="m-1 text-green-400" />
+        )}
+      </div>
+    </div>
+  )
+}
+
+export const TransactionsTableDetailsTeleporterSteps = ({
+  tx,
+  address
+}: {
+  tx: MergedTransaction
+  address: Address | undefined
+}) => {
+  const { isTestnet: isTestnetTx } = isNetwork(tx.childChainId)
+
+  const l2TxID = tx.l1ToL2MsgData?.l2TxID
+  const isFirstRetryableSucceeded =
+    typeof l2TxID !== 'undefined' &&
+    typeof tx.l2ToL3MsgData?.l2ForwarderRetryableTxID === 'undefined'
+  const l2ChainId = tx.l2ToL3MsgData?.l2ChainId
+  const isFirstRetryableFailed = firstRetryableLegRequiresRedeem(tx)
+  const _l2ForwarderRequiresRedeem = l2ForwarderRetryableRequiresRedeem(tx)
+
+  const isFirstRetryableResolved =
+    isFirstRetryableSucceeded || isFirstRetryableFailed
+
+  const secondRetryableResolved =
+    l2L3RetryableRequiresRedeem(tx) ||
+    typeof tx.l2ToL3MsgData?.l3TxID !== 'undefined'
+
+  const firstTransactionText = useMemo(() => {
+    if (isFirstRetryableFailed) {
+      const l2NetworkName = getNetworkName(getChainIdForRedeemingRetryable(tx))
+
+      return (
+        <div>
+          Transaction failed on {l2NetworkName}. You have 7 days to retry. After
+          that, the tx is no longer recoverable.
+          {/* if we detect we will have 2 redemptions in the first leg of teleport, explain it to users */}
+          {_l2ForwarderRequiresRedeem && (
+            <TeleportMiddleStepFailureExplanationNote tx={tx} />
+          )}
+        </div>
+      )
+    }
+
+    // if l2ChainId is present
+    if (l2ChainId) {
+      return `Funds arrived on ${getNetworkName(l2ChainId)}`
+    }
+
+    // till the time we don't have information for l2ChainId
+    return `Funds arrived on intermediate chain`
+  }, [tx, isFirstRetryableFailed, l2ChainId, _l2ForwarderRequiresRedeem])
+
+  const firstRetryableRedeemButton = useMemo(
+    () => (
+      <TransactionsTableRowAction
+        type="deposits"
+        isError={true}
+        tx={tx}
+        address={address}
+      />
+    ),
+    [tx, address]
+  )
+
+  const firstTransactionExternalLink = useMemo(
+    () =>
+      l2TxID &&
+      l2ChainId && (
+        <ExternalLink href={`${getExplorerUrl(l2ChainId)}/tx/${l2TxID}`}>
+          <ArrowTopRightOnSquareIcon height={12} />
+        </ExternalLink>
+      ),
+    [l2TxID, l2ChainId]
+  )
+
+  const firstTransactionActionItem = isFirstRetryableFailed
+    ? firstRetryableRedeemButton
+    : firstTransactionExternalLink
+
+  const firstRetryableWaitingDuration = useMemo(
```

Comment:
I'm not sure if `useMemo` here, and in `secondRetryableWaitingDuration`, would do more harm than good.